### ISSUE TITLE
Running code from a plain *.sc file is slow, but fast if put into a function

### ISSUE THREAD

**Issue by cheapsolutionarchitect** [VirtusLab/scala-cli#2462]
**Version(s)**
Scala CLI version: 1.0.2, JDK 11

**Describe the bug**

Running code that relies on the CPU is very slow when directly run from a plain *.sc file. It runs substantially faster after  it was put into a function.

**To Reproduce**

Rename the file [slow.txt](https://github.com/VirtusLab/scala-cli/files/12896752/slow.txt) to slow.sc and run it with "scala-cli -S 2.13.10 slow.sc". This will take minutes.

Rename the second file  [fast.txt](https://github.com/VirtusLab/scala-cli/files/12896751/fast.txt) to fast.sc and run it with "scala-cli -S 2.13.10 fast.sc". This will take a few seconds (13s on my machine).

**Expected behaviour**

The code from slow.txt should run as fast as the code from fast.txt, without the need to wrap it into a function.

**Comment by SethTisue** [comment]
offhand, it looks like the script wrapper isn't set up to avoid the usual JVM pitfall where code run only once in the body of a constructor doesn't get JITted?

**Comment by JD557** [comment]
When trying to reproduce this I noticed that this seems to only happen with Java 8 and 11, not 17 and up (I didn't try any non-LTS version).

**Comment by JD557** [comment]
From a quick analysis the bytecode seems quite different on the for loop, from a `public static final void $anonfun$findLimitsBruteForce$1(int, int, int, scala.runtime.IntRef, scala.runtime.ObjectRef, int);` to a `public static final void $anonfun$new$1(int)`

The problem seems to go away if the constants are marked as `final`, as that changes the first layer of the loop from (a similar thing happens on the other layers):

```
  public static final void $anonfun$new$1(int);
    Code:
       0: getstatic     #62                 // Field scala/runtime/RichInt$.MODULE$:Lscala/runtime/RichInt$;
       3: getstatic     #67                 // Field scala/Predef$.MODULE$:Lscala/Predef$;
       6: iconst_1
       7: invokevirtual #71                 // Method scala/Predef$.intWrapper:(I)I
      10: getstatic     #33                 // Field MODULE$:Lunwrapped$;
      13: invokevirtual #189                // Method b:()I
      16: invokevirtual #78                 // Method scala/runtime/RichInt$.to$extension:(II)Lscala/collection/immutable/Range$Inclusive;
      19: iload_0
      20: invokedynamic #195,  0            // InvokeDynamic #2:apply$mcVI$sp:(I)Lscala/runtime/java8/JFunction1$mcVI$sp;
      25: invokevirtual #100                // Method scala/collection/immutable/Range$Inclusive.foreach$mVc$sp:(Lscala/Function1;)V
      28: return
```

to

```
  public static final void $anonfun$new$1(int);
    Code:
       0: getstatic     #50                 // Field scala/runtime/RichInt$.MODULE$:Lscala/runtime/RichInt$;
       3: getstatic     #55                 // Field scala/Predef$.MODULE$:Lscala/Predef$;
       6: iconst_1
       7: invokevirtual #59                 // Method scala/Predef$.intWrapper:(I)I
      10: sipush        300
      13: invokevirtual #63                 // Method scala/runtime/RichInt$.to$extension:(II)Lscala/collection/immutable/Range$Inclusive;
      16: iload_0
      17: invokedynamic #179,  0            // InvokeDynamic #2:apply$mcVI$sp:(I)Lscala/runtime/java8/JFunction1$mcVI$sp;
      22: invokevirtual #85                 // Method scala/collection/immutable/Range$Inclusive.foreach$mVc$sp:(Lscala/Function1;)V
      25: return
```

Notice that the `invokevirtual` for `b()` disappeared in favor of a `sipush 300` (the constant value for b)

I wonder if this is a compiler bug, as I thought all fields in an object should be automatically marked as final... but I guess that's not the case 🤔 - looks like the fields are final (e.g. `private static final int max;`), but not the accessors (`public int max();`).

**Comment by JD557** [comment]
I created an issue in the compiler, but looks like this is the expected behavior.

As such, maybe it would make sense for the scala-cli wrapper to be slightly different? Namely wrap everything into a `main` function instead of just wrapping it into an object?

**Comment by som-snytt** [comment]
These are ancient considerations for scala 2 repl and runner. `object` constructors are run under class init `<clinit>`, causing non-jit performance. The runner wraps code in an anon class in the main method, so you get class-constructor semantics. The other choice for snippets is constructor vs block, where overloading is not allowed in local blocks.

**Comment by MaciejG604** [comment]
The performance problem is resolved when using jvm 17 as pointed out by @JD557 and also by using the new wrapper variation with `--delayed-init`. It's proposed in #2535

**Comment by MaciejG604** [comment]
Fixed by #2556, no flag will be needed, the new mechanism will be the default for Scala 2
